libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs

```diff
@@ -615,30 +617,89 @@ genValidUMapNonEmpty = do
   (rdPairs, ptrs, sPools, dReps) <- genValidTuplesNonEmpty
   pure $ unify rdPairs ptrs sPools dReps
 
-uniformSubset ::
+uniformSubSet ::
   (StatefulGen g m, Ord k) =>
   -- | Size of the subset. If supplied will be clamped to @[0, Set.size s]@ interval,
   -- otherwise will be generated randomly.
   Maybe Int ->
   Set k ->
   g ->
   m (Set k)
-uniformSubset mSubSetSize inputSet gen = do
+uniformSubSet mSubSetSize inputSet gen = do
   subSetSize <- case mSubSetSize of
     Nothing -> uniformRM (0, Set.size inputSet) gen
     Just n -> pure $ max 0 $ min (Set.size inputSet) n
-  go inputSet Set.empty subSetSize
+  if subSetSize < Set.size inputSet `div` 2
+    then
+      goAdd inputSet Set.empty subSetSize
+    else
+      goDelete inputSet (Set.size inputSet - subSetSize)
   where
-    go !s !acc !i
+    -- Constructing a new Set is faster when less then a half of original Set will be used
+    goAdd !s !acc !i
       | i <= 0 = pure acc
       | otherwise = do
           ix <- uniformRM (0, Set.size s - 1) gen
-          go (Set.insert (Set.elemAt ix s) acc) (Set.deleteAt ix s) (i - 1)
+          goAdd (Set.deleteAt ix s) (Set.insert (Set.elemAt ix s) acc) (i - 1)
+    -- Deleting is faster when more items need to be retained in the Set
+    goDelete !acc !i
+      | i <= 0 = pure acc
+      | otherwise = do
+          ix <- uniformRM (0, Set.size acc - 1) gen
+          goDelete (Set.deleteAt ix acc) (i - 1)
+
+uniformSubMap ::
+  (StatefulGen g m, Ord k) =>
+  -- | Size of the subMap. If supplied will be clamped to @[0, Map.size s]@ interval,
+  -- otherwise will be generated randomly.
+  Maybe Int ->
+  Map k v ->
+  g ->
+  m (Map k v)
+uniformSubMap mSubMapSize inputMap gen = do
+  subMapSize <- case mSubMapSize of
+    Nothing -> uniformRM (0, Map.size inputMap) gen
+    Just n -> pure $ max 0 $ min (Map.size inputMap) n
+  if subMapSize < Map.size inputMap `div` 2
+    then
+      -- Constructing a new Map is faster when less then a half of original Map will be used
+      uniformSubMapElems Map.insert (Just subMapSize) inputMap gen
+    else
+      -- Deleting is faster when more items need to be retained in the Map
+      goDelete inputMap (Map.size inputMap - subMapSize)
+  where
+    goDelete !acc !i
+      | i <= 0 = pure acc
+      | otherwise = do
+          ix <- uniformRM (0, Map.size acc - 1) gen
+          goDelete (Map.deleteAt ix acc) (i - 1)
+
+uniformSubMapElems ::
+  (StatefulGen g m, Monoid f) =>
+  (k -> v -> f -> f) ->
+  -- | Size of the subMap. If supplied will be clamped to @[0, Map.size s]@ interval,
+  -- otherwise will be generated randomly.
+  Maybe Int ->
+  Map k v ->
+  g ->
+  m f
+uniformSubMapElems insert mSubMapSize inputMap gen = do
+  subMapSize <- case mSubMapSize of
+    Nothing -> uniformRM (0, Map.size inputMap) gen
+    Just n -> pure $ max 0 $ min (Map.size inputMap) n
```

Comment:
Would it make sense to extract this logic of determining the size of a the subset/submap, since it appears three times?

```
getSubSize ::StatefulGen g m =>
  Maybe Int ->
  Int ->
  g ->
  m  Int
getSubSize mSubSetSize inputSize gen = do
  case mSubSetSize of
    Nothing -> uniformRM (0, inputSize) gen
    Just n -> pure $ max 0 $ min inputSize n
```

And then this would become: 

```suggestion
subMapSize <- getSubSize mSubMapSize (Map.size inputMap) gen
```

Similar for the other two functions, `uniformSubMap` and `uniformSubSet`

Reply:
Very nice. I really like this suggestion. Applied.